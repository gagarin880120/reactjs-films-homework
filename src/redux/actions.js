function searchResultsAction(results) {
  return {
      type: 'SEARCH',
      results: results
  };
}

function genresAction(genres) {
  return {
      type: 'GENRES',
      genres: genres
  };
}

export { searchResultsAction, genresAction };
