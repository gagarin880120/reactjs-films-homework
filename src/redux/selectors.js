import getGenresString from '../helpers/helpers';

const searchResultsSelector = (state) => {
  state.searchResults.map((v) => {
    const obj = v;
    obj.genres = getGenresString(v.genre_ids, state.genres);
    return { ...obj };
  });
  return state.searchResults;
};

const modalSelector = (state) => state.isModalOpen;

const trailerSelector = (state) => state.trailerURL;


export { searchResultsSelector, modalSelector, trailerSelector };
