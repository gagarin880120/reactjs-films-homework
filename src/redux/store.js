import { createStore, applyMiddleware } from 'redux';
import thunk from 'redux-thunk';

const initialState = {
  searchResults: [],
  genres: [],
  isModalOpen: false,
  trailerURL: '',
  isTrailerLoaded: false,
};

export function reducer(state = initialState, action) {
  switch (action.type) {
    case 'SEARCH':
      return { ...state, searchResults: action.results };
    case 'GENRES':
      return { ...state, genres: action.genres };
    case 'MODAL':
      return { ...state, isModalOpen: action.isModalOpen };
    case 'TRAILER':
      return { ...state, trailerURL: action.trailerURL };
    case 'IS_TRAILER_LOADED':
      return { ...state, isTrailerLoaded: action.isTrailerLoaded };
    default:
      return state;
  }
}

export const store = createStore(reducer, applyMiddleware(thunk));
