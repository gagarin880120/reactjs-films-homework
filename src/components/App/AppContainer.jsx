import React, { useEffect } from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import { getGenres, getMovies } from '../../redux/actions';
import { genresSelector } from '../../redux/selectors';
import getURL from '../../utils/utils';
import App from './App';
import Spinner from '../Spinner/Spinner';
import styles from './App.module.scss';

export function AppContainer({ onAppLoad, genres }) {
  useEffect(() => {
    if (!genres) {
      onAppLoad();
    }
  }, []);

  return (
    genres
      ? <App /> : (
        <div className={styles.loading}>
          <Spinner />
        </div>
      )
  );
}

export const mapStateToProps = (state) => ({
  genres: genresSelector(state),
});

export const mapDispatchToProps = (dispatch) => ({
  onAppLoad() {
    dispatch(getGenres());
    dispatch(getMovies(getURL('trending'), 1));
  },
});

AppContainer.propTypes = {
  onAppLoad: PropTypes.func,
  genres: PropTypes.instanceOf(Array),
};

AppContainer.defaultProps = {
  onAppLoad: null,
  genres: [],
};

export default connect(mapStateToProps, mapDispatchToProps)(AppContainer);
