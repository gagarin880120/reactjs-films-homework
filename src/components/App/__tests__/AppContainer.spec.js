import ShallowRenderer from 'react-test-renderer/shallow';
import React from 'react';
import { AppContainer, mapStateToProps, mapDispatchToProps } from '../AppContainer';
import { getGenres, getMovies } from '../../../redux/actions';

jest.mock('../../../redux/actions',() => ({
  getGenres: jest.fn().mockReturnValue('getGenresAction'),
  getMovies: jest.fn().mockReturnValue('getMoviesAction'),
}))

describe('AppContainer component', () => {
  test('should render without crashing', () => {
    const renderer = new ShallowRenderer();
    renderer.render(<AppContainer />);
    const result = renderer.getRenderOutput();
    expect(result).toMatchSnapshot();
  });

  describe('mapStateToProps', () => {
    test('should return the right value', () => {

      const initialState = {
        genres: [{id: 28, name: 'Action'}]
      };

      expect(mapStateToProps(initialState).genres).toEqual([{id: 28, name: 'Action'}]);
    });
  });

  describe('mapDispatchToProps', () => {
    afterEach(() => {
      jest.clearAllMocks();
    })

    test('onAppLoad should dispatch getGenres actions', () => {
      const dispatch = jest.fn();
      const {onAppLoad} = mapDispatchToProps(dispatch);

      onAppLoad();

      expect(getGenres).toHaveBeenCalled();
      expect(getMovies).toHaveBeenCalled();
    })
  })

});
