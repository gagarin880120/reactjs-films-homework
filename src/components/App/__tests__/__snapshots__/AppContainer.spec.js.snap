// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AppContainer component should render without crashing 1`] = `
<BrowserRouter>
  <App
    currentAPIRequest=""
    genres={Array []}
    onAppLoad={null}
    onReload={null}
  />
</BrowserRouter>
`;
